xmppserver/src/main/java/org/jivesoftware/openfire/muc/spi/LocalMUCRoom.java

```diff
@@ -931,28 +937,36 @@ public void leaveRoom(OccupantLeftEvent event) {
      * @param originator true if this JVM is the one that originated the event.
      */
     private void removeOccupantRole(MUCRole leaveRole, boolean originator) {
-        JID userAddress = leaveRole.getUserAddress();
-        // Notify the user that he/she is no longer in the room
-        leaveRole.destroy();
-        // Update the tables of occupants based on the bare and full JID
-        JID bareJID = userAddress.asBareJID();
-
-        String nickname = leaveRole.getNickname();
-        List<MUCRole> occupants = occupantsByNickname.get(nickname.toLowerCase());
-        if (occupants != null) {
-            occupants.remove(leaveRole);
-            if (occupants.isEmpty()) {
-                occupantsByNickname.remove(nickname.toLowerCase());
+        JID userAddress;
+        String nickname;
+        lock.writeLock().lock();
+        try {
+            userAddress = leaveRole.getUserAddress();
+            // Notify the user that he/she is no longer in the room
+            leaveRole.destroy();
+            // Update the tables of occupants based on the bare and full JID
+            JID bareJID = userAddress.asBareJID();
+
+            nickname = leaveRole.getNickname();
+            List<MUCRole> occupants = occupantsByNickname.get(nickname.toLowerCase());
```

Comment:
Lock doesn't need to be taken until this point.

Reply:
done